Network/Curl/Url.hs

```diff
@@ -0,0 +1,43 @@
+-- | Bindings to curl/urlapi.h
+module Network.Curl.Url (isValidUrl) where
+
+import Foreign (IntPtr (..))
+import Foreign.C (CInt (..), CUInt (..))
+import Foreign.C.String (CString, withCString)
+import GHC.IO (unsafePerformIO)
+
+-- | Ask cURL whether it would consider the given string a valid URL.
+--
+-- cURL attempts to adhere to RFC 3986.
+isValidUrl :: String -> Bool
+isValidUrl s = unsafePerformIO $ do
+  -- Allocate a CURLU struct and store a pointer to it
+  handle <- curl_url
+  code <- withCString s $ \str -> do
+    -- cURL parses the string when we set the whole URL, and returns
+    -- a code if the URL is bad.
+    curl_url_set
+      handle -- the CURLU*
+      0      -- CURLUPART_URL
+      str    -- the string (which gets copied)
+      0      -- dont pass any CURLU flags
+  -- Cleanup: only the CURLU handle needs to be freed, because the
+  -- CString we made earlier gets copied into it (so we can deallocate it
+  -- after acquiring the code, which holds no references to the string)
+  curl_url_cleanup handle
+  -- The string is a valid URL if cURL has no complaints about it; that is,
+  -- the `curl_url_set` we did earlier returned `CURLUE_OK` (== 0)
+  return (code == 0)
+
+type Flags = CUInt
+
+-- functions imported from `curl/urlapi.h`
+
+foreign import ccall
+  "curl_url" curl_url :: IO IntPtr
+
+foreign import ccall
+  "curl_url_cleanup" curl_url_cleanup :: IntPtr -> IO ()
+
+foreign import ccall
```

Comment:
Can you make the `safety` explicit in these definitions? I think these can also be `unsafe` as they do not call back into Haskell.